### PR TITLE
Multiple gems in git repo bugfix

### PR DESCRIPTION
**Pull request by joelnordell**
# Description:

This change is an important fix for when a Gemfile specifies multiple gems that are contained in one git repository (where each gem is in a subdirectory within the repo) and they all have native extensions.

Without this change, in certain situations bundler would only build one of the native extensions but not the others.

______________

The fix is to separate the extensions build path for each gem inside the repository, so that the `gem.build_complete` file exists independently for each separate gem.

______________

# Tasks:

- [X] Describe the problem / feature
- [X] Write tests
- [X] Write code to solve the problem
- [x] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

______________

This PR replaces https://github.com/rubygems/rubygems/pull/3435 and adds a test.

### CODE REVIEW

**Comment by joelnordell** [comment]
@crugg -- FYI, here is the rspec test for the bundler fix.

File: bundler/spec/install/gems/native_extensions_spec.rb [hunk 1]
```diff
@@ -88,6 +88,62 @@
...
+
+    bundle! "config set build.c_extension_one --with-c_extension_one=one"
+    bundle! "config set build.c_extension_two --with-c_extension_two=two"
+
+    # 1st time, require only one gem -- only one of the extensions gets built.
+    install_gemfile! <<-G
+      gem "c_extension_one", :git => #{lib_path("gems").to_s.dump}
+    G
+
+    expect(out).not_to include("extconf.rb failed")
```
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/native_extensions_spec.rb:133]: I know this is probably copied from other specs in this file that do the same, but this line is not necessary, since the `install_gemfile!` will fail earlier in case of an error when building the extension.

File: bundler/spec/install/gems/native_extensions_spec.rb [hunk 2]
```diff
@@ -88,6 +88,62 @@
...
+
+    # 2nd time, require both gems -- we need both extensions to be built now.
+    install_gemfile! <<-G
+      gem "c_extension_one", :git => #{lib_path("gems").to_s.dump}
+      gem "c_extension_two", :git => #{lib_path("gems").to_s.dump}
+    G
+
+    expect(out).not_to include("extconf.rb failed")
```
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/native_extensions_spec.rb:141]: Same as the other comment here.

File: bundler/spec/install/gems/native_extensions_spec.rb (hunk 1 above)
Comment (joelnordell) [reply bundler/spec/install/gems/native_extensions_spec.rb:133]: Yep, that was taken from the other adjacent tests to be consistent with them. I'm fine with removing it since the final expectation can only pass if everything succeeds with the build - should I remove it from all the other tests in that file too?

**Comment by joelnordell** [comment]
> While reviewing it, I noticed a warning being printed when bundle installing this kind of setups (not sure if you noticed it). Anyways, since you've been going over this area of the code, I appreciate if you can have a look at it, it's #3479.

Sure, I'll take a look.

File: bundler/spec/install/gems/native_extensions_spec.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply bundler/spec/install/gems/native_extensions_spec.rb:133]: I'm doing that in #3479, so not necessary!

**Comment by joelnordell** [comment]
@deivid-rodriguez two of the checks here failed but seem to be unrelated to the changes in this PR. Is there any way to re-run those checks?

**Comment by deivid-rodriguez** [comment]
Can you rebase this PR to make sure it runs of top of the current master? Just in case.

File: bundler/spec/install/gems/native_extensions_spec.rb [hunk 3]
```diff
@@ -87,6 +87,58 @@
     expect(out).to eq("true")
   end
 
+  it "installs from git when multiple gems share one repository" do
```
Comment (deivid-rodriguez) [inline bundler/spec/install/gems/native_extensions_spec.rb:90]: To make the wording of the spec more exact, I'd say:

```suggestion
  it "installs correctly from git when multiple gems with extensions share one repository" do
```

Just pointing out that the test is about extensions, and adding "correctly" to try to emphasize that the install won't actually fail, but will install "incorrectly" (extensions artifacts overwriting each other).

**Comment by joelnordell** [comment]
> I finished to review this. Just a little tweak to the spec wording, and also to anwer [@segiddins' question in the original PR](https://github.com/rubygems/bundler/pull/6934#issuecomment-460010015):
> 
> > Will this require rebuilding all existing extensions?
> 
> I don't think so. This PR should be fully "backwards compatible" in that sense, since the target folder name is preserved _unless_ there's a conflicting situation like the one in this issue. Please @joelnordell correct me if I'm wrong.

That is correct. To demonstrate this, here are some results showing where bundler builds native extensions for two of the test cases (one gem in one git repo, the previously existing test; and multiple gems in one repo, the new test) - both with and without this change:

---

### One gem in one repo (existing test case - most typical scenario):
Running from `master`:
```
$ rspec spec/install/gems/native_extensions_spec.rb:49
$ find tmp/1/gems/system/bundler/gems/ -name gem.build_complete
tmp/1/gems/system/bundler/gems/extensions/x86_64-linux/2.7.0/c_extension-1.0-f5582e00bb79/gem.build_complete
```

Running from this branch:
```
$ rspec spec/install/gems/native_extensions_spec.rb:49
$ find tmp/1/gems/system/bundler/gems/ -name gem.build_complete
tmp/1/gems/system/bundler/gems/extensions/x86_64-linux/2.7.0/c_extension-1.0-ed2ed6b79d5a/gem.build_complete
```

You can see here that the path where the extension is built is unchanged (aside from the git commit hash, which is different every time the test runs).

---

### Multiple gems in one repo (new test case - scenario that previously did not work):
Running from `master` (with the new test cherry-picked):
```sh
$ rspec spec/install/gems/native_extensions_spec.rb:138 # NOTE the test fails here
$ find tmp/1/gems/system/bundler/gems/ -name gem.build_complete
tmp/1/gems/system/bundler/gems/gems-bb0b90940942/extensions/x86_64-linux/2.7.0/gems-bb0b90940942/gem.build_complete
```

Running from this branch:
```
$ rspec spec/install/gems/native_extensions_spec.rb:138
$ find tmp/1/gems/system/bundler/gems/ -name gem.build_complete
tmp/1/gems/system/bundler/gems/gems-63da39124084/extensions/x86_64-linux/2.7.0/gems-63da39124084-c_extension_one/gem.build_complete
tmp/1/gems/system/bundler/gems/gems-63da39124084/extensions/x86_64-linux/2.7.0/gems-63da39124084-c_extension_two/gem.build_complete
```

In this case, however, with this change the extensions are built in different paths so that `gem.build_complete` exists independently for each extension.

---

The bottom line is this change is backwards compatible for the most common case of a repository containing only one gem which is located in the root of the repository.

For the less common case of a gem being contained in a subdirectory of a repository (regardless of how many there are), when upgrading bundler this would result in the previously-built extension no longer being recognized as having been built, and it would then get rebuilt.

**Comment by joelnordell** [comment]
@deivid-rodriguez - this should be ready to go now - updated the test description and squashed to one commit.
